### ISSUE TITLE
Installing Octo tool occassionally fails

### ISSUE THREAD

**Issue by mcasperson** [OctopusDeploy/OctoTFS#115]
From https://help.octopus.com/t/unable-to-find-octo-exe-but-its-just-been-installed/21525.

This is the log from a failed attempt to download and use a new version of the Octo client. This is an intermittent issue.

```
2018-10-17T02:39:52.4646373Z ==============================================================================
2018-10-17T02:39:52.4646506Z Task : Package Application
2018-10-17T02:39:52.4646614Z Description : Package your application into a NuPkg or Zip file.
2018-10-17T02:39:52.4646728Z Version : 3.0.180
2018-10-17T02:39:52.4646824Z Author : Octopus Deploy
2018-10-17T02:39:52.4646920Z Help : Version: 3.0.180. More Information
2018-10-17T02:39:52.4647045Z ==============================================================================
2018-10-17T02:39:53.6391699Z Checking local tool cache
2018-10-17T02:39:53.6450057Z Downloading: https://download.octopusdeploy.com/octopus-tools/4.41.5/OctopusTools.4.41.5.portable.zip
2018-10-17T02:39:53.6965884Z Extracting archive
2018-10-17T02:39:53.7012757Z [command]/usr/bin/unzip /home/vsts/work/_temp/3ceab605-2ec0-40e7-9e37-95336c0b6699
2018-10-17T02:39:53.7701306Z Archive: /home/vsts/work/_temp/3ceab605-2ec0-40e7-9e37-95336c0b6699
2018-10-17T02:39:53.8142009Z inflating: Autofac.dll
2018-10-17T02:39:53.8147976Z inflating: Markdown.dll
2018-10-17T02:39:53.8152898Z inflating: Microsoft.Extensions.Caching.Abstractions.dll
2018-10-17T02:39:53.8158186Z inflating: Microsoft.Extensions.Caching.Memory.dll
2018-10-17T02:39:53.8164461Z inflating: Microsoft.Extensions.DependencyInjection.Abstractions.dll
2018-10-17T02:39:53.8168173Z inflating: Microsoft.Extensions.Options.dll
2018-10-17T02:39:53.8173040Z inflating: Microsoft.Extensions.Primitives.dll
2018-10-17T02:39:53.8225342Z inflating: Newtonsoft.Json.dll
2018-10-17T02:39:53.8230932Z inflating: NuGet.Common.dll
2018-10-17T02:39:53.8244212Z inflating: NuGet.Frameworks.dll
2018-10-17T02:39:53.8248270Z inflating: NuGet.Packaging.Core.dll
2018-10-17T02:39:53.8251122Z inflating: NuGet.Packaging.Core.Types.dll
2018-10-17T02:39:53.8270207Z inflating: NuGet.Packaging.dll
2018-10-17T02:39:53.8275941Z inflating: NuGet.Versioning.dll
2018-10-17T02:39:53.8277190Z inflating: Octo
2018-10-17T02:39:53.8278965Z inflating: Octo.cmd
2018-10-17T02:39:53.8285129Z inflating: Octo.deps.json
2018-10-17T02:39:53.8288116Z inflating: Octo.dll
2018-10-17T02:39:53.8289701Z inflating: Octo.pdb
2018-10-17T02:39:53.8291254Z inflating: Octo.runtimeconfig.json
2018-10-17T02:39:53.8299113Z inflating: Octodiff.dll
2018-10-17T02:39:53.8345195Z inflating: Octopus.Cli.dll
2018-10-17T02:39:53.8356488Z inflating: Octopus.Cli.pdb
2018-10-17T02:39:53.8421518Z inflating: Octopus.Client.dll
2018-10-17T02:39:53.8421669Z inflating: Octopus.Client.dll.config
2018-10-17T02:39:53.8439963Z inflating: Octopus.Client.pdb
2018-10-17T02:39:53.8452581Z inflating: Octopus.Client.xml
2018-10-17T02:39:53.8462564Z inflating: Octostache.dll
2018-10-17T02:39:53.8476234Z inflating: Serilog.dll
2018-10-17T02:39:53.8478655Z inflating: Serilog.Sinks.ColoredConsole.dll
2018-10-17T02:39:53.8480776Z inflating: Serilog.Sinks.Trace.dll
2018-10-17T02:39:53.8488960Z inflating: Sprache.dll
2018-10-17T02:39:53.8540818Z Caching tool: Octo 4.41.5 x64
2018-10-17T02:39:53.8890298Z Looking for /opt/hostedtoolcache/Octo/4.41.5/x64/*Octo.dll
2018-10-17T02:39:53.9013471Z Prepending PATH environment variable with directory: /opt/hostedtoolcache/Octo/4.41.5/x64
2018-10-17T02:39:53.9118139Z ##[error]Error: Unable to find or install octo.
2018-10-17T02:39:53.9131477Z ##[error]Failed to execute octo pack command. Unable to find or install octo.
```

**Comment by mcasperson** [comment]
https://github.com/OctopusDeploy/OctoTFS/pull/99 will address this by falling back to an embedded copy of the octo cli if the download fails.

**Comment by mcasperson** [comment]
The octo cli tool is now embedded into the plugin, and if a download fails the embedded version will be used as a fallback.

**Comment by ArturDorochowicz** [comment]
@mcasperson Are you guys sure #99 is related to this issue?

I have seen this problem for months with various 3.* versions, most recently yesterday with version 3.0.189. I can't tell if #99 is present in this version (is it?).

The problem is quite infrequent and non deterministic, which is consistent with a network issue. OTOH, as you can see from the log, there's no network error reported and actually the archive seems to be extracted with no problems.

I have only seen this occur on Linux machines, never on Windows.

**Comment by mcasperson** [comment]
Hi @ArturDorochowicz ,

The PR for #99 is available in 3.0.189. While it won't resolve network issues, it provides the `Octopus tools installer` step that will fall back to an embedded copy of the Octopus CLI if a version can not be downloaded, which prevents deployments from failing with the `Error: Unable to find or install octo`.

![image](https://user-images.githubusercontent.com/160104/48869599-8a876900-ee29-11e8-9284-db8bf4c9e9f6.png)

**Comment by ArturDorochowicz** [comment]
Ok, if I understand correctly, the fallback to embedded copy only happens in `Octopus tools installer` and not in other tasks (even though they may also download Octopus client).

I suppose I will need to add the `Octopus tools installer` task to my release pipelines and observe the behavior.

But besides that, what I'm trying to point out is that, judging by the log in the OP (and I have seen essentially the same in all instances of failures in my environment), there's no network error there and it is confirmed by the archive expanding successfully.
